dali/test/python/test_RN50_external_source_parallel.py

```diff
@@ -0,0 +1,363 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import argparse
+import time
+import cv2
+import numpy as np
+
+import nvidia.dali as dali
+import nvidia.dali.types as types
+from nvidia.dali.plugin.base_iterator import LastBatchPolicy
+
+from test_utils import AverageMeter
+
+
+class FilesDataSet:
+
+    def __init__(self, data_path):
+        files = self.get_files_list(data_path)
+        self.data_path = data_path
+        self.classes_no = len(files)
+        self.files_map = {}
+
+        counter = 0
+        for class_no, (_, samples) in enumerate(files):
+            for sample in samples:
+                self.files_map[counter] = (sample, class_no)
+                counter += 1
+        self.size = counter
+
+    @classmethod
+    def get_files_list(cls, data_path):
+        dirs = [
+            (dir_name, [
+                file_path for file_path
+                in map(lambda name: os.path.join(dir_path, name), os.listdir(dir_path))
+                if os.path.isfile(file_path)
+            ])
+            for (dir_name, dir_path)
+            in map(lambda name: (name, os.path.join(data_path, name)), os.listdir(data_path))
+            if os.path.isdir(dir_path)
+        ]
+        dirs.sort(key=lambda dir_files: dir_files[0])
+        for _, files in dirs:
+            files.sort()
+        return dirs
+
+    def __len__(self):
+        return self.size
+
+    def __getitem__(self, i):
+        if i >= self.size:
+            raise StopIteration
+        return self.files_map[i]
+
+
+class SampleLoader(object):
+
+    DATA_SET = FilesDataSet
+
+    def __init__(self, data_path):
+        self.data_path = data_path
+        self.data_set = self.create_data_set()
+
+    def __getattr__(self, key):
+        if key == 'data_set':
+            self._data_set = self._data_set or self.create_data_set()
+            self.__dict__['data_set'] = self._data_set
+            return self._data_set
+        raise AttributeError
+
+    def create_data_set(self):
+        return self.DATA_SET(data_path=self.data_path)
+
+    def read_file(self, file_path):
+        with open(file_path, 'rb') as f:
+            return f.read()
+
+    def __call__(self, sample_info):
+        file_path, class_no = self.data_set[sample_info.idx_in_epoch]
+        return self.read_file(file_path), np.array([class_no])
+
+
+class ShuffledFilesDataSet(FilesDataSet):
+
+        def __init__(self, data_path):
+            super().__init__(data_path)
+            self.perm = np.random.permutation(self.size)
+
+        def __getitem__(self, i):
+            if i >= self.size:
+                raise StopIteration
```

Comment:
Maybe you can permeate here again (you need to set seed manually to predictable value to make sure that each shard has the same permutation).

Reply:
Even it doesn't change anything here we can refer to it when anyone asks.